File: app/tasks/strategy_runner.py
import asyncio

import orjson

from app.cache.redis import (
    redis_client
)

from app.strategy_engine.strategy_manager import (
    StrategyManager
)

from app.core.logger import logger


async def main():

    logger.info(
        "Starting Strategy Runner..."
    )

    pubsub = redis_client.pubsub()

    await pubsub.subscribe(
        "market_ticks"
    )

    logger.info(
        "Subscribed to market_ticks"
    )

    strategy_manager = StrategyManager()

    async def sync_loop():
        while True:
            try:
                await strategy_manager.sync_strategies()
            except Exception as e:
                logger.error(f"Strategy Sync Error: {e}")
            await asyncio.sleep(5)

    asyncio.create_task(sync_loop())

    while True:

        message = await pubsub.get_message(
            ignore_subscribe_messages=True,
            timeout=1.0
        )

        if message is None:

            await asyncio.sleep(
                0.01
            )

            continue

        try:

            tick = orjson.loads(
                message["data"]
            )

            logger.info(
                f"Received Tick: {tick}"
            )

            signal = (
                await strategy_manager.process_market_data(
                    symbol=tick["symbol"],
                    price=tick["price"]
                )
            )

            if signal:

                logger.info(
                    f"Generated Signal: {signal}"
                )

        except Exception as e:

            logger.error(
                f"Strategy Runner Error: {e}"
            )


if __name__ == "__main__":

    asyncio.run(main())

File: backend/src/workers/strategyRunner.ts
import { createRedisClient } from "../core/redis";
import { StrategyManager } from "../trading/strategies/strategyManager";
import { logger } from "../core/logger";

async function main() {
  logger.info("Starting Strategy Runner...");
  const pubsub = createRedisClient();

  await pubsub.subscribe("market_ticks");
  logger.info("Subscribed to market_ticks");

  const strategyManager = new StrategyManager();

  // Sync strategies immediately on startup
  try {
    await strategyManager.syncStrategies();
  } catch (err: any) {
    logger.error(`Initial Strategy Sync Error: ${err.message}`);
  }

  // Set up 5s sync interval
  const syncInterval = setInterval(async () => {
    try {
      await strategyManager.syncStrategies();
    } catch (e: any) {
      logger.error(`Strategy Sync Error: ${e.message}`);
    }
  }, 5000);

  pubsub.on("message", async (channel, message) => {
    if (channel !== "market_ticks") return;

    try {
      const tick = JSON.parse(message);
      logger.info(`Received Tick: ${message}`);
      await strategyManager.processMarketData(tick.symbol, tick.price);
    } catch (e: any) {
      logger.error(`Strategy Runner Error: ${e.message}`);
    }
  });

  return () => {
    clearInterval(syncInterval);
    pubsub.disconnect();
  };
}

async function runWithRecovery() {
  while (true) {
    let cleanup: (() => void) | undefined;
    try {
      cleanup = await main();
      // Keep alive
      await new Promise(() => {});
    } catch (e: any) {
      if (cleanup) cleanup();
      logger.error(`Strategy Runner crashed: ${e.message}`);
      if (e.stack) logger.error(e.stack);
      logger.info("Restarting Strategy Runner in 5 seconds...");
      await new Promise((resolve) => setTimeout(resolve, 5000));
    }
  }
}

runWithRecovery();
